### ISSUE TITLE
Napari viewer closes unexpectetly when pressing the 3D button

### ISSUE THREAD

**Issue by sebi06** [napari/napari#678]
## 🐛 Bug

Hi guys,

I use the following code to automatically display a 3D stack with 2 Channels and 10 Timepoints in Napari. This works fine and the viewer opens up just fine.
When I then press the 3D button the viewer is gone and I get a long error stack: See below:
Any help is welcome.

Sebi

![image](https://user-images.githubusercontent.com/3833249/68606163-216f8a00-04ae-11ea-906c-726f3195e4fe.png)

```
def show_napari(array, metadata, verbose=True):

    import napari

    with napari.gui_qt():

        # create scalefcator with all ones
        scalefactors = [1] * len(array.shape)
        
        # initialize the napari viewer
        viewer = napari.Viewer()
            
        if metadata['ImageType'] == 'czi':

            # find position of dimensions
            posZ = metadata['Axes'].find('Z')
            posC = metadata['Axes'].find('C')
            posT = metadata['Axes'].find('T')

            # get the scalefactors from the metadata
            scalef = get_scalefactor(metadata)
            scalefactors[posZ] = scalef['zx']

            if verbose:
                print('Dim PosT : ', posT)
                print('Dim PosZ : ', posZ)
                print('Dim PosC : ', posC)
                print('Scale Factors XYZ: ', scalefactors)
            
            # add all channels as layers
            for ch in range(metadata['SizeC']):
                chname = metadata['Channels'][ch]

                # cut out channel
                channel = array.take(ch, axis=posC)
                print(channel.shape)

                # actually show the image array
                print('Adding Channel: ', chname)
                viewer.add_image(channel, name=chname, scale=scalefactors)
```
Error message:

```
(1, 10, 2, 15, 256, 256, 1)
BTCZYX0
(1, 10, 2, 15, 256, 256)
Dim PosT :  1
Dim PosZ :  3
Dim PosC :  2
Scale Factors XYZ:  [1, 1, 1, 3.516, 1, 1]
(1, 10, 15, 256, 256)
Adding Channel:  AF555
(1, 10, 15, 256, 256)
Adding Channel:  AF488
WARNING: Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\_qt\qt_viewer_buttons.py", line 173, in <lambda>
    lambda state=self: self.change_ndisplay(state)
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\_qt\qt_viewer_buttons.py", line 178, in change_ndisplay
    self.viewer.dims.ndisplay = 3
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\components\dims.py", line 205, in ndisplay
    self.events.ndisplay()
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\util\event.py", line 508, in __call__
    self._invoke_callback(cb, event)
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\util\event.py", line 529, in _invoke_callback
    cb_event=(cb, event),
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\util\event.py", line 523, in _invoke_callback
    cb(event)
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\components\viewer_model.py", line 89, in <lambda>
    self.dims.events.ndisplay.connect(lambda e: self._update_layers())
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\components\viewer_model.py", line 1018, in _update_layers
    layer.dims.ndisplay = self.dims.ndisplay
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\components\dims.py", line 205, in ndisplay
    self.events.ndisplay()
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\util\event.py", line 508, in __call__
    self._invoke_callback(cb, event)
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\util\event.py", line 529, in _invoke_callback
    cb_event=(cb, event),
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\util\event.py", line 523, in _invoke_callback
    cb(event)
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\layers\base\base.py", line 174, in <lambda>
    self.dims.events.ndisplay.connect(lambda e: self._update_dims())
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\layers\base\base.py", line 372, in _update_dims
    self._set_view_slice()
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\layers\image\image.py", line 502, in _set_view_slice
    self.events.set_data()
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\util\event.py", line 508, in __call__
    self._invoke_callback(cb, event)
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\util\event.py", line 529, in _invoke_callback
    cb_event=(cb, event),
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\util\event.py", line 523, in _invoke_callback
    cb(event)
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\_vispy\vispy_base_layer.py", line 46, in <lambda>
    self.layer.events.set_data.connect(lambda e: self._on_data_change())
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\_vispy\vispy_image_layer.py", line 59, in _on_data_change
    self._on_display_change()
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\_vispy\vispy_image_layer.py", line 48, in _on_display_change
    self.reset()
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\_vispy\vispy_image_layer.py", line 216, in reset
    self._reset_base()
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\_vispy\vispy_base_layer.py", line 165, in _reset_base
    self._on_scale_change()
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\_vispy\vispy_image_layer.py", line 123, in _on_scale_change
    self.layer.position = self._transform_position(self._position)
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\_vispy\vispy_base_layer.py", line 153, in _transform_position
    transform.map(list(position))[: len(self.layer.dims.displayed)]
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\vispy\visuals\transforms\chain.py", line 148, in map
    coords = tr.map(coords)
  File "<decorator-gen-4>", line 2, in imap
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\vispy\visuals\transforms\_util.py", line 111, in arg_to_vec4
    arg = as_vec4(arg)
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\vispy\visuals\transforms\_util.py", line 81, in as_vec4
    % obj.shape)
TypeError: not all arguments converted during string formatting
WARNING:vispy:Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\_qt\qt_viewer_buttons.py", line 173, in <lambda>
    lambda state=self: self.change_ndisplay(state)
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\_qt\qt_viewer_buttons.py", line 178, in change_ndisplay
    self.viewer.dims.ndisplay = 3
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\components\dims.py", line 205, in ndisplay
    self.events.ndisplay()
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\util\event.py", line 508, in __call__
    self._invoke_callback(cb, event)
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\util\event.py", line 529, in _invoke_callback
    cb_event=(cb, event),
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\util\event.py", line 523, in _invoke_callback
    cb(event)
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\components\viewer_model.py", line 89, in <lambda>
    self.dims.events.ndisplay.connect(lambda e: self._update_layers())
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\components\viewer_model.py", line 1018, in _update_layers
    layer.dims.ndisplay = self.dims.ndisplay
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\components\dims.py", line 205, in ndisplay
    self.events.ndisplay()
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\util\event.py", line 508, in __call__
    self._invoke_callback(cb, event)
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\util\event.py", line 529, in _invoke_callback
    cb_event=(cb, event),
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\util\event.py", line 523, in _invoke_callback
    cb(event)
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\layers\base\base.py", line 174, in <lambda>
    self.dims.events.ndisplay.connect(lambda e: self._update_dims())
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\layers\base\base.py", line 372, in _update_dims
    self._set_view_slice()
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\layers\image\image.py", line 502, in _set_view_slice
    self.events.set_data()
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\util\event.py", line 508, in __call__
    self._invoke_callback(cb, event)
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\util\event.py", line 529, in _invoke_callback
    cb_event=(cb, event),
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\util\event.py", line 523, in _invoke_callback
    cb(event)
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\_vispy\vispy_base_layer.py", line 46, in <lambda>
    self.layer.events.set_data.connect(lambda e: self._on_data_change())
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\_vispy\vispy_image_layer.py", line 59, in _on_data_change
    self._on_display_change()
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\_vispy\vispy_image_layer.py", line 48, in _on_display_change
    self.reset()
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\_vispy\vispy_image_layer.py", line 216, in reset
    self._reset_base()
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\_vispy\vispy_base_layer.py", line 165, in _reset_base
    self._on_scale_change()
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\_vispy\vispy_image_layer.py", line 123, in _on_scale_change
    self.layer.position = self._transform_position(self._position)
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\napari\_vispy\vispy_base_layer.py", line 153, in _transform_position
    transform.map(list(position))[: len(self.layer.dims.displayed)]
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\vispy\visuals\transforms\chain.py", line 148, in map
    coords = tr.map(coords)
  File "<decorator-gen-4>", line 2, in imap
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\vispy\visuals\transforms\_util.py", line 111, in arg_to_vec4
    arg = as_vec4(arg)
  File "C:\ProgramData\Anaconda3\envs\imageanalysis\lib\site-packages\vispy\visuals\transforms\_util.py", line 81, in as_vec4
    % obj.shape)
TypeError: not all arguments converted during string formatting
```

**Comment by sofroniewn** [comment]
Hi @sebi06 - great catch. I was able to reproduce with a minimal example:

```python
viewer = napari.view_image(np.random.random((2, 10, 20, 30, 25)))
viewer.dims.ndisplay = 3
```

We were not initializing one of our variables properly. Sorry about that. This is being fixed in #680 (still under review) and a test has been added to prevent it appearing again.

Please let us know if the fix in #680 works for you

**Comment by sebi06** [comment]
I refreshed my napari repository and installed via

`pip install -e .`

And the problem is gone and works as expected now. So I can remove my workaround. Thank you.
